akka-typed/src/main/scala/akka/typed/cluster/sharding/ClusterSharding.scala

```diff
@@ -0,0 +1,259 @@
+   * the message will be `unhandled`, i.e. posted as `Unhandled` messages on the event stream
+   */
+  def entityId(message: E): String
+
+  /**
+   * Extract the message to send to the entity from an incoming `message`.
+   * Note that the extracted message does not have to be the same as the incoming
+   * message to support wrapping in message envelope that is unwrapped before
+   * sending to the entity actor.
+   *
+   * If the returned value is `null`, and the entity isn't running yet the entity will be started
+   * but no message will be delivered to it.
+   */
+  def entityMessage(message: E): A // TODO "unwrapMessage" is how I'd call it?
+
+  /**
+   * Extract the entity id from an incoming `message`. Only messages that passed the [[#entityId]]
+   * function will be used as input to this function.
+   */
+  def shardId(message: E): String
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using envelopes to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+// TODO not sure"DefaultMessageExtractor", not sure that name is good; perhaps explain what it does in name?
+final class DefaultMessageExtractor[A](maxNumberOfShards: Int) extends ShardingMessageExtractor[ShardingEnvelope[A], A] {
+  def entityId(envelope: ShardingEnvelope[A]): String = envelope.entityId
+  def entityMessage(envelope: ShardingEnvelope[A]): A = envelope.message
+  def shardId(envelope: ShardingEnvelope[A]): String = (math.abs(envelope.entityId.hashCode) % maxNumberOfShards).toString
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using a property of the message to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+abstract class DefaultNoEnvelopeMessageExtractor[A](maxNumberOfShards: Int) extends ShardingMessageExtractor[A, A] {
+  def entityMessage(message: A) = message
+  def shardId(message: A) = {
+    val id = entityId(message)
+    if (id != null) (math.abs(id.hashCode) % maxNumberOfShards).toString
+    else null
+  }
+}
+
+object ClusterSharding extends ExtensionId[ClusterSharding] {
+
+  /** Scala API */
+  override def createExtension(system: ActorSystem[_]): ClusterSharding =
+    new ClusterShardingImpl(system)
+
+  /** Java API */
+  def get(system: ActorSystem[_]): ClusterSharding = apply(system)
+}
+
+final class ClusterShardingImpl(system: ActorSystem[_]) extends ClusterSharding {
+  require(system.isInstanceOf[ActorSystemAdapter[_]], "only adapted actor systems can be used for cluster features")
```

Comment:
"only untyped actor system" ?

Reply:
It's not really "untyped" that we want, we want the adapted one... so maybe "only adapted untyped actor systems..."

Reply:
ok